x-pack/plugin/ccr/src/test/java/org/elasticsearch/xpack/ccr/action/ShardFollowTaskReplicationTests.java

```diff
@@ -281,6 +287,59 @@ public void testRetryBulkShardOperations() throws Exception {
         }
     }
 
+    /**
+     * This is not a following test; it's mostly copied from RecoveryDuringReplicationTests#testAddNewReplicas.
+     * This test verifies the MSU optimization which is currently only enabled on {@link FollowingEngine}.
+     */
+    public void testAddNewReplicasWithFollowingEngine() throws Exception {
+        Settings settings = Settings.builder()
+            .put(IndexSettings.INDEX_SOFT_DELETES_SETTING.getKey(), true)
+            .put(CcrSettings.CCR_FOLLOWING_INDEX_SETTING.getKey(), true)
+            .build();
+        IndexMetaData indexMetaData = buildIndexMetaData(between(0, 1), settings, indexMapping);
+        try (ReplicationGroup shards = new ReplicationGroup(indexMetaData) {
+            @Override
+            protected EngineFactory getEngineFactory(ShardRouting routing) {
+                if (routing.primary()) {
+                    return new InternalEngineFactory(); // use the internal engine so we can index directly
```

Comment:
why can't we use the standard following engine and supply ops in the same way we do for real following? the goal is to check that everything works together like it should.

Reply:
@bleskes Yes, we can but I chose not to because we have an integration test does the same thing (https://github.com/elastic/elasticsearch/pull/34474/files#diff-d36012b5317805e8b3ff77b77dbaa5ecR672). I am fine to update this. WDYT?

Reply:
ok, I didn't want to fully spec out the testing, but I would have ideally loved to see a randomized non multi threaded test using replication groups that create all kind of scenarios (like the one you have caught). The multi threaded test can stay in integration land and make sure that if we miss a scenario something will find it. Does that make sense?

Reply:
@bleskes Thanks for the explanation 👍. I will update this test.